### ISSUE TITLE
LinalgFuse failed with the new fusion heuristics

### ISSUE THREAD

**Issue by pzread** [iree-org/iree#10228]
With the https://github.com/pzread/iree/tree/fusion_heuristics (based on https://github.com/iree-org/iree/pull/10177), the sample below failed the LinalgFuse

With the command `iree-compile --iree-mlir-to-vm-bytecode-module --iree-input-type=none --iree-hal-target-backends=llvm-cpu --iree-llvm-target-triple=x86_64-linux-gnu --iree-llvm-target-cpu-features=host --mlir-print-ir-after-all ./multiuser.mlir`

```mlir
func.func @main(%arg0 : tensor<6x3xf32>) -> tensor<6x3xf32> {
  %cst = arith.constant 0.000000e+00 : f32
  %cst_0 = arith.constant 0xFF800000 : f32
  %0 = linalg.init_tensor [6] : tensor<6xf32>
  %1 = linalg.fill ins(%cst_0 : f32) outs(%0 : tensor<6xf32>) -> tensor<6xf32>
  %2 = linalg.generic {
      indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>],
      iterator_types = ["parallel", "reduction"]}
      ins(%arg0 : tensor<6x3xf32>) outs(%1 : tensor<6xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %9 = arith.maxf %arg1, %arg2 : f33
      linalg.yield %9 : f32
    } -> tensor<6xf32>
  %3 = linalg.init_tensor [6, 3] : tensor<6x3xf32>
  %4 = linalg.generic {
       indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>],
       iterator_types = ["parallel", "parallel"]}
       ins(%arg0, %2 : tensor<6x3xf32>, tensor<6xf32>)
       outs(%3 : tensor<6x3xf32>) {
    ^bb0(%arg1: f32, %arg2: f32, %arg3: f32):
      %9 = arith.subf %arg1, %arg2 : f32
      %10 = math.exp %9 : f32
      linalg.yield %10 : f32
    } -> tensor<6x3xf32>
  %5 = linalg.fill ins(%cst : f32) outs(%0 : tensor<6xf32>) -> tensor<6xf32>
  %6 = linalg.generic {
      indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>],
      iterator_types = ["parallel", "reduction"]}
      ins(%4 : tensor<6x3xf32>) outs(%5 : tensor<6xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %9 = arith.addf %arg1, %arg2 : f32
      linalg.yield %9 : f32
    } -> tensor<6xf32>
  %7 = linalg.generic {
      indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>],
      iterator_types = ["parallel", "parallel"]}
      ins(%4, %6 : tensor<6x3xf32>, tensor<6xf32>)
      outs(%3 : tensor<6x3xf32>) {
    ^bb0(%arg1: f32, %arg2: f32, %arg3: f32):
      %9 = arith.divf %arg1, %arg2 : f32
      linalg.yield %9 : f32
    } -> tensor<6x3xf32>
  return %7 : tensor<6x3xf32>
}
```

**Comment by pzread** [comment]
@MaheshRavishankar FYI

**Comment by MaheshRavishankar** [comment]
isnt this softmax?

**Comment by pzread** [comment]
Yes, it is. I'm not sure why it doesn't work on my side.

**Comment by pzread** [comment]
The dump of the failed LinalgFuse
```mlir
// -----// IR Dump After LinalgFuse Failed (linalg-fuse) //----- //
mlir-asm-printer: Verifying operation: func.func
func.func @main_dispatch_0_generic_6x3() {
  %c6 = arith.constant 6 : index
  %c0 = arith.constant 0 : index
  %cst = arith.constant 0.000000e+00 : f32
  %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:6x3xf32>
  %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:6xf32>
  %2 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<writeonly:6x3xf32>
  %workgroup_id_x = hal.interface.workgroup.id[0] : index
  %workgroup_count_x = hal.interface.workgroup.count[0] : index
  %3 = affine.apply affine_map<()[s0] -> (s0 * 6)>()[%workgroup_id_x]
  %4 = affine.apply affine_map<()[s0] -> (s0 * 6)>()[%workgroup_count_x]
  scf.for %arg0 = %3 to %c6 step %4 {
    %5 = flow.dispatch.tensor.load %2, offsets = [%arg0, 0], sizes = [6, 3], strides = [1, 1] : !flow.dispatch.tensor<writeonly:6x3xf32> -> tensor<6x3xf32>
    %6 = flow.dispatch.tensor.load %0, offsets = [%arg0, 0], sizes = [6, 3], strides = [1, 1] : !flow.dispatch.tensor<readonly:6x3xf32> -> tensor<6x3xf32>
    %7 = flow.dispatch.tensor.load %1, offsets = [%arg0], sizes = [6], strides = [1] : !flow.dispatch.tensor<readonly:6xf32> -> tensor<6xf32>
    %8 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%6 : tensor<6x3xf32>) outs(%7 : tensor<6xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %16 = arith.maxf %arg1, %arg2 : f32
      linalg.yield %16 : f32
    } -> tensor<6xf32>
    %9 = linalg.init_tensor [6] : tensor<6xf32>
    %10 = linalg.fill ins(%cst : f32) outs(%9 : tensor<6xf32>) -> tensor<6xf32>
    %11:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%6, %8 : tensor<6x3xf32>, tensor<6xf32>) outs(%5, %10 : tensor<6x3xf32>, tensor<6xf32>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[6, 0], [6, 0], [0, 16]]>} {
    ^bb0(%arg1: f32, %arg2: f32, %arg3: f32, %arg4: f32):
      %16 = arith.subf %arg1, %arg2 : f32
      %17 = math.exp %16 : f32
      %18 = arith.addf %17, %arg4 : f32
      linalg.yield %17, %18 : f32, f32
    } -> (tensor<6x3xf32>, tensor<6xf32>)
    %12 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%6 : tensor<6x3xf32>) outs(%7 : tensor<6xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %16 = arith.maxf %arg1, %arg2 : f32
      linalg.yield %16 : f32
    } -> tensor<6xf32>
    %13 = linalg.fill ins(%cst : f32) outs(%9 : tensor<6xf32>) -> tensor<6xf32>
    %14 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%6, %12 : tensor<6x3xf32>, tensor<6xf32>) outs(%13 : tensor<6xf32>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[6, 0], [6, 0], [0, 16]]>} {
    ^bb0(%arg1: f32, %arg2: f32, %arg3: f32):
      %16 = arith.subf %arg1, %arg2 : f32
      %17 = math.exp %16 : f32
      %18 = arith.addf %17, %arg3 : f32
      linalg.yield %18 : f32
    } -> tensor<6xf32>
    %15 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%14 : tensor<6xf32>) outs(%11#0 : tensor<6x3xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):
      %16 = arith.divf %arg2, %arg1 : f32
      linalg.yield %16 : f32
    } -> tensor<6x3xf32>
    flow.dispatch.tensor.store %15, %2, offsets = [%arg0, 0], sizes = [6, 3], strides = [1, 1] : tensor<6x3xf32> -> !flow.dispatch.tensor<writeonly:6x3xf32>
  }
  return
}
```

**Comment by pzread** [comment]
Looks like there are two `lowering_config` for some reason... So the LinalgFuse couldn't find the root op.

**Comment by MaheshRavishankar** [comment]
Should be fixed by https://github.com/iree-org/iree/commit/cf5a5d5dd507ceb90538fb89fae4dc124d8bdad7 . I am rebasing #10177  on top of main now. Ill add some tests to it as well...

**Comment by pzread** [comment]
I think my version is the same as cf5a5d5 + #10177, I'll look deeper. Let me know if you have a workable branch

**Comment by MaheshRavishankar** [comment]
rebasing and trying out few examples... Ill update then

**Comment by MaheshRavishankar** [comment]
Actually I see whats happening. `TileAndDistributeToWorkgroups` is the cause here. Before that the IR is

```
hal.executable.variant public @embedded_elf_x86_64, target = <"llvm-cpu", "embedded-elf-x86_64", {cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}> {
  hal.executable.export public @_softmax_dispatch_0_generic_12x128x128 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>) attributes {translation_info = #iree_codegen.translation_info<CPUDoubleTilingExpert>} {
  ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index):
    %x, %y, %z = flow.dispatch.default_workgroup_count %arg1, %arg2, %arg3
    hal.return %x, %y, %z : index, index, index
  }
  builtin.module {
    func.func @_softmax_dispatch_0_generic_12x128x128() {
      %c0 = arith.constant 0 : index
      %cst = arith.constant 0.000000e+00 : f32
      %cst_0 = arith.constant 1.000000e+00 : f32
      %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:12x128x128xf32>
      %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:12x128xf32>
      %2 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<writeonly:12x128x128xf32>
      %3 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [12, 128, 128], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:12x128x128xf32> -> tensor<12x128x128xf32>
      %4 = flow.dispatch.tensor.load %1, offsets = [0, 0], sizes = [12, 128], strides = [1, 1] : !flow.dispatch.tensor<readonly:12x128xf32> -> tensor<12x128xf32>
      %5 = linalg.init_tensor [12, 128, 128] : tensor<12x128x128xf32>
      %6 = linalg.init_tensor [12, 128] : tensor<12x128xf32>
      %7 = linalg.fill ins(%cst : f32) outs(%6 : tensor<12x128xf32>) -> tensor<12x128xf32>
      %8 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%3 : tensor<12x128x128xf32>) outs(%4 : tensor<12x128xf32>) {
      ^bb0(%arg0: f32, %arg1: f32):
        %11 = arith.maxf %arg0, %arg1 : f32
        linalg.yield %11 : f32
      } -> tensor<12x128xf32>
      %9:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%3, %8 : tensor<12x128x128xf32>, tensor<12x128xf32>) outs(%5, %7 : tensor<12x128x128xf32>, tensor<12x128xf32>) attrs =  {lowering_config = #ir\
ee_codegen.lowering_config<tile_sizes = [[4, 32, 0], [1, 4, 0], [0, 0, 4]]>} {
      ^bb0(%arg0: f32, %arg1: f32, %arg2: f32, %arg3: f32):
        %11 = arith.subf %arg0, %arg1 : f32
        %12 = math.exp %11 : f32
        %13 = arith.addf %12, %arg3 : f32
        linalg.yield %12, %13 : f32, f32
      } -> (tensor<12x128x128xf32>, tensor<12x128xf32>)
      %10 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%9#0, %9#1 : tensor<12x128x128xf32>, tensor<12x128xf32>) outs(%5 : tensor<12x128x128xf32>) {
      ^bb0(%arg0: f32, %arg1: f32, %arg2: f32):
        %11 = arith.divf %cst_0, %arg1 : f32
        %12 = arith.mulf %arg0, %11 : f32
        linalg.yield %12 : f32
      } -> tensor<12x128x128xf32>
      flow.dispatch.tensor.store %10, %2, offsets = [0, 0, 0], sizes = [12, 128, 128], strides = [1, 1, 1] : tensor<12x128x128xf32> -> !flow.dispatch.tensor<writeonly:12x128x128xf32>
      return
    }
  }
}
```

That has only three generic ops. After tile and distribute some ops get replicated

```
      scf.for %arg0 = %3 to %c12 step %4 {
        %5 = affine.apply affine_map<()[s0] -> (s0 * 32)>()[%workgroup_id_x]
        %6 = affine.apply affine_map<()[s0] -> (s0 * 32)>()[%workgroup_count_x]
        scf.for %arg1 = %5 to %c128 step %6 {
          %7 = flow.dispatch.tensor.load %0, offsets = [%arg0, %arg1, 0], sizes = [%c4, %c32, 128], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:12x128x128xf32> -> tensor<?x?x128xf32>
          %8 = flow.dispatch.tensor.load %0, offsets = [%arg0, %arg1, 0], sizes = [%c4, %c32, 128], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:12x128x128xf32> -> tensor<?x?x128xf32>
          %9 = flow.dispatch.tensor.load %1, offsets = [%arg0, %arg1], sizes = [%c4, %c32], strides = [1, 1] : !flow.dispatch.tensor<readonly:12x128xf32> -> tensor<?x?xf32>
          %10 = tensor.cast %9 : tensor<?x?xf32> to tensor<4x32xf32>
          %11 = tensor.cast %8 : tensor<?x?x128xf32> to tensor<4x32x128xf32>
          %12 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%11 : tensor<4x32x128xf32>) outs(%10 : tensor<4x32xf32>) {
          ^bb0(%arg2: f32, %arg3: f32):
            %31 = arith.maxf %arg2, %arg3 : f32
            linalg.yield %31 : f32
          } -> tensor<4x32xf32>
          %13 = linalg.init_tensor [4, 32] : tensor<4x32xf32>
          %14 = linalg.fill ins(%cst : f32) outs(%13 : tensor<4x32xf32>) -> tensor<4x32xf32>
          %15 = linalg.init_tensor [4, 32, 128] : tensor<4x32x128xf32>
          %16 = tensor.cast %7 : tensor<?x?x128xf32> to tensor<4x32x128xf32>
          %17:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%16, %12 : tensor<4x32x128xf32>, tensor<4x32xf32>) outs(%15, %14 : tensor<4x32x128xf32>, tensor<4x32xf32>) attrs =  {lowering_config = #i\
ree_codegen.lowering_config<tile_sizes = [[4, 32, 0], [1, 4, 0], [0, 0, 4]]>} {
          ^bb0(%arg2: f32, %arg3: f32, %arg4: f32, %arg5: f32):
            %31 = arith.subf %arg2, %arg3 : f32
            %32 = math.exp %31 : f32
            %33 = arith.addf %32, %arg5 : f32
            linalg.yield %32, %33 : f32, f32
          } -> (tensor<4x32x128xf32>, tensor<4x32xf32>)
          %18 = flow.dispatch.tensor.load %0, offsets = [%arg0, %arg1, 0], sizes = [%c4, %c32, 128], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:12x128x128xf32> -> tensor<?x?x128xf32>
          %19 = flow.dispatch.tensor.load %0, offsets = [%arg0, %arg1, 0], sizes = [%c4, %c32, 128], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:12x128x128xf32> -> tensor<?x?x128xf32>
          %20 = flow.dispatch.tensor.load %1, offsets = [%arg0, %arg1], sizes = [%c4, %c32], strides = [1, 1] : !flow.dispatch.tensor<readonly:12x128xf32> -> tensor<?x?xf32>
          %21 = tensor.cast %20 : tensor<?x?xf32> to tensor<4x32xf32>
          %22 = tensor.cast %19 : tensor<?x?x128xf32> to tensor<4x32x128xf32>
          %23 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%22 : tensor<4x32x128xf32>) outs(%21 : tensor<4x32xf32>) {
          ^bb0(%arg2: f32, %arg3: f32):
            %31 = arith.maxf %arg2, %arg3 : f32
            linalg.yield %31 : f32
          } -> tensor<4x32xf32>
          %24 = linalg.init_tensor [4, 32] : tensor<4x32xf32>
          %25 = linalg.fill ins(%cst : f32) outs(%24 : tensor<4x32xf32>) -> tensor<4x32xf32>
          %26 = tensor.cast %18 : tensor<?x?x128xf32> to tensor<4x32x128xf32>
          %27 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%26, %23 : tensor<4x32x128xf32>, tensor<4x32xf32>) outs(%25 : tensor<4x32xf32>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[4, 32, 0], [1, 4, 0], [0, 0,\
 4]]>} {
          ^bb0(%arg2: f32, %arg3: f32, %arg4: f32):
            %31 = arith.subf %arg2, %arg3 : f32
            %32 = math.exp %31 : f32
            %33 = arith.addf %32, %arg4 : f32
            linalg.yield %33 : f32
          } -> tensor<4x32xf32>
          %28 = linalg.init_tensor [4, 32, 128] : tensor<4x32x128xf32>
          %29 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%17#0, %27 : tensor<4x32x128xf32>, tensor<4x32xf32>) outs(%28 : tensor<4x32x128xf32>) {
          ^bb0(%arg2: f32, %arg3: f32, %arg4: f32):
            %31 = arith.divf %cst_0, %arg3 : f32
            %32 = arith.mulf %arg2, %31 : f32
            linalg.yield %32 : f32
          } -> tensor<4x32x128xf32>
          %30 = tensor.cast %29 : tensor<4x32x128xf32> to tensor<?x?x128xf32>
          flow.dispatch.tensor.store %30, %2, offsets = [%arg0, %arg1, 0], sizes = [%c4, %c32, 128], strides = [1, 1, 1] : tensor<?x?x128xf32> -> !flow.dispatch.tensor<writeonly:12x128x128xf32>
        }
      }
```

Need to adapt the logic of fusion in `TileAndDistributeToWorkgroups` to only fuse an op once

**Comment by pzread** [comment]
I just realized the same thing. I can try to fix this if you haven't started.

**Comment by MaheshRavishankar** [comment]
Fix might be a bit involved though. Definitely worth digging into it for you to understand the issue. We can discuss solutions. I have some ideas, and I can give you pointers

**Comment by MaheshRavishankar** [comment]
The `TileAndDistributeToWorkgroups` implements a version of this https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp

Upstream there is a pattern for Tiling and a pattern for TileAndFuse. In IREE I thought only using the Tiling pattern is enough cause we don't have to deal with `scf.for` with destructive updates. We can use `scf.for` with `flow.dispatch.tensor.load/store` and avoid having to deal with `iter_args`. 

Here though, we have multiple uses from the same producer. Fusing along each use individually might not work. We have to fuse all uses at once and check that all uses from the producer to the consumer can be handled. This means we need to probably introduce a `TileAndFuse` pattern in IREe as well similar to upstream. 

TL;DR the fix is a bit involved. Let me take a stab at it over the weekend. We can sync up next week on what's happening here. Worth for you to look at what the upstream code is doing and what we do in IREE today.

**Comment by pzread** [comment]
I was testing a method to fix the duplicate issue in TileAndDistributeToWorkgroups
https://github.com/iree-org/iree/commit/12b2115aea14bffc391e49e721f8d13e4be32e55

It can now tile all generic ops only once. But the same issue occurs in LinalgStrategyTileAndFusePass
```mlir
// -----// IR Dump After LinalgStrategyTileAndFusePass (linalg-strategy-tile-and-fuse-pass) //----- //
func.func @main_dispatch_0_generic_6x3() {
  %c6 = arith.constant 6 : index
  %c0 = arith.constant 0 : index
  %cst = arith.constant 0xFF800000 : f32
  %cst_0 = arith.constant 0.000000e+00 : f32
  %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<readonly:6x3xf32>
  %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : !flow.dispatch.tensor<writeonly:6x3xf32>
  %workgroup_id_x = hal.interface.workgroup.id[0] : index
  %workgroup_count_x = hal.interface.workgroup.count[0] : index
  %2 = affine.apply affine_map<()[s0] -> (s0 * 6)>()[%workgroup_id_x]
  %3 = affine.apply affine_map<()[s0] -> (s0 * 6)>()[%workgroup_count_x]
  scf.for %arg0 = %2 to %c6 step %3 {
    %4 = flow.dispatch.tensor.load %1, offsets = [%arg0, 0], sizes = [6, 3], strides = [1, 1] : !flow.dispatch.tensor<writeonly:6x3xf32> -> tensor<6x3xf32>
    %5 = flow.dispatch.tensor.load %0, offsets = [%arg0, 0], sizes = [6, 3], strides = [1, 1] : !flow.dispatch.tensor<readonly:6x3xf32> -> tensor<6x3xf32>
    %6 = linalg.init_tensor [6] : tensor<6xf32>
    %7 = scf.for %arg1 = %c0 to %c6 step %c6 iter_args(%arg2 = %4) -> (tensor<6x3xf32>) {
      %8 = tensor.extract_slice %5[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> to tensor<6x3xf32>
      %9 = tensor.extract_slice %5[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> to tensor<6x3xf32>
      %10 = tensor.extract_slice %6[%arg1] [6] [1] : tensor<6xf32> to tensor<6xf32>
      %11 = linalg.fill {__internal_linalg_transform__ = "1"} ins(%cst : f32) outs(%10 : tensor<6xf32>) -> tensor<6xf32>
      %12 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%9 : tensor<6x3xf32>) outs(%11 : tensor<6xf32>) attrs =  {__internal_linalg_transform__ = "1"} {
      ^bb0(%arg3: f32, %arg4: f32):
        %28 = arith.maxf %arg3, %arg4 : f32
        linalg.yield %28 : f32
      } -> tensor<6xf32>
      %13 = tensor.extract_slice %4[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> to tensor<6x3xf32>
      %14 = tensor.extract_slice %6[%arg1] [6] [1] : tensor<6xf32> to tensor<6xf32>
      %15 = linalg.fill {__internal_linalg_transform__ = "1"} ins(%cst_0 : f32) outs(%14 : tensor<6xf32>) -> tensor<6xf32>
      %16:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%8, %12 : tensor<6x3xf32>, tensor<6xf32>) outs(%13, %15 : tensor<6x3xf32>, tensor<6xf32>) attrs =  {__internal_linalg_transform__ = "1", lowering_config = #iree_codegen.lowering_config<tile_sizes = [[6, 0], [6, 0], [0, 16]]>} {
      ^bb0(%arg3: f32, %arg4: f32, %arg5: f32, %arg6: f32):
        %28 = arith.subf %arg3, %arg4 : f32
        %29 = math.exp %28 : f32
        %30 = arith.addf %29, %arg6 : f32
        linalg.yield %29, %30 : f32, f32
      } -> (tensor<6x3xf32>, tensor<6xf32>)
      %17 = tensor.extract_slice %arg2[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> to tensor<6x3xf32>
      %18 = tensor.extract_slice %5[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> to tensor<6x3xf32>
      %19 = tensor.extract_slice %5[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> to tensor<6x3xf32>
      %20 = tensor.extract_slice %6[%arg1] [6] [1] : tensor<6xf32> to tensor<6xf32>
      %21 = linalg.fill {__internal_linalg_transform__ = "1"} ins(%cst : f32) outs(%20 : tensor<6xf32>) -> tensor<6xf32>
      %22 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%19 : tensor<6x3xf32>) outs(%21 : tensor<6xf32>) attrs =  {__internal_linalg_transform__ = "1"} {
      ^bb0(%arg3: f32, %arg4: f32):
        %28 = arith.maxf %arg3, %arg4 : f32
        linalg.yield %28 : f32
      } -> tensor<6xf32>
      %23 = tensor.extract_slice %6[%arg1] [6] [1] : tensor<6xf32> to tensor<6xf32>
      %24 = linalg.fill {__internal_linalg_transform__ = "1"} ins(%cst_0 : f32) outs(%23 : tensor<6xf32>) -> tensor<6xf32>
      %25:2 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%18, %22 : tensor<6x3xf32>, tensor<6xf32>) outs(%17, %24 : tensor<6x3xf32>, tensor<6xf32>) attrs =  {__internal_linalg_transform__ = "1", lowering_config = #iree_codegen.lowering_config<tile_sizes = [[6, 0], [6, 0], [0, 16]]>} {
      ^bb0(%arg3: f32, %arg4: f32, %arg5: f32, %arg6: f32):
        %28 = arith.subf %arg3, %arg4 : f32
        %29 = math.exp %28 : f32
        %30 = arith.addf %29, %arg6 : f32
        linalg.yield %29, %30 : f32, f32
      } -> (tensor<6x3xf32>, tensor<6xf32>)
      %26 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%16#1 : tensor<6xf32>) outs(%25#0 : tensor<6x3xf32>) attrs =  {__internal_linalg_transform__ = "1"} {
      ^bb0(%arg3: f32, %arg4: f32):
        %28 = arith.divf %arg4, %arg3 : f32
        linalg.yield %28 : f32
      } -> tensor<6x3xf32>
      %27 = tensor.insert_slice %26 into %arg2[%arg1, 0] [6, 3] [1, 1] : tensor<6x3xf32> into tensor<6x3xf32>
      scf.yield %27 : tensor<6x3xf32>
    }
    flow.dispatch.tensor.store %7, %1, offsets = [%arg0, 0], sizes = [6, 3], strides = [1, 1] : tensor<6x3xf32> -> !flow.dispatch.tensor<writeonly:6x3xf32>
  }
  return
}
```

**Comment by MaheshRavishankar** [comment]
Hey @pzread . Skimmed through what you have, and that is about what right. I think for the use cases in IREE this works. But I was fairly certain the upstream tile + fuse would also have this issue. Basically [this](https://github.com/llvm/llvm-project/blob/6732896bbf0c525d79fbfef7b4b23ef0a56f7862/mlir/lib/Dialect/Linalg/Transforms/FusionOnTensors.cpp#L423) method needs to be deprecated cause it cannot be fixed easily for this use case. I have been working on something in IREE, that I can move upstream ( https://github.com/MaheshRavishankar/iree/commit/1e1236b9ad41438d2de572bbf3dbb6702a583ae0 ).

The only difference here is that there is no `llvm::MapVector` to map from original operands to fused operands. That is tricky to manage in a general DAG where the same operand could have multiple uses with different indexing maps. This was basically tried upstream and removed (the code removed in this commit has some of the same logic, but also accounts for accesses through different indexing maps https://github.com/llvm/llvm-project/commit/485190df95f98c51c3f4a4ab4db96127cdc9ce78).
Still, all the complication it is accounting for is not needed in IREE for the level that the issue is at. The commit I pointed to is something that I will upstream as well. Ill send out that PR for review shortly (need to add comments and lit tests)

**Comment by pzread** [comment]
The method you implemented right now seems to be more complete. But I feel like there might be a bug in the worklist loop https://github.com/MaheshRavishankar/iree/commit/1e1236b9ad41438d2de572bbf3dbb6702a583ae0#diff-a2963394002544132b964ee72bc2aa37266aab0c96c6e470960f058aa01953d1R572

Since when looping the worklist, it keeps creating more slices for the tiled producers, the candidate list might be incomplete and we will tile a producer too early (some of its uses haven't been tiled and created slices). This is why I chose DFS instead of BFS in my version.

For example, given A -> C and A -> B -> C:

1. First tile A, then the candidate list has the slices of B and C
2. The loop decides to tile C first, but it doesn't include the use from B, because B hasn't been tiled.
3. Then the loop tiles B, now a new slice of C is created, but the C has been tiled.

**Comment by MaheshRavishankar** [comment]
You are probably right about DFS vs BFS. Ill change it.
